grails-app/services/org/pih/warehouse/dashboard/IndicatorDataService.groovy

```diff
@@ -1111,6 +1112,74 @@ class IndicatorDataService {
         return graphData.toJson()
     }
 
+    Map getBackdatedShipmentsChart(String query, String locationId, Integer monthsLimit) {
+        Map<String, Integer> data = [
+                '2 days': 0,
+                '3 days': 0,
+                '4 days': 0,
+                '5 days': 0,
+                '6 days': 0,
+                '7 days': 0,
+                '7+ days': 0,
+        ]
+        Date timeLimit = LocalDate.now().minusMonths(monthsLimit).toDate()
+        List queryData = dataService.executeQuery(query, [
+                locationId: locationId,
+                timeLimit: timeLimit,
+                daysOffset: Holders.config.openboxes.dashboard.backdate.daysOffset
+        ])
+
+        queryData.each {
+            data[it[0]] = it[1]
+        }
+
+        List<IndicatorDatasets> datasets = [
+                new IndicatorDatasets('Backdated shipments', data*.value.toList())
+        ]
+        IndicatorData indicatorData = new IndicatorData(datasets, data*.key.toList())
+        GraphData graphData = new GraphData(indicatorData)
+
+        return graphData.toJson()
+    }
+
+    @Cacheable(value = "dashboardCache", key = { "getBackdatedOutboundShipments-${locationId}${monthsLimit}" })
+    Map getBackdatedOutboundShipmentsData(String locationId, Integer monthsLimit) {
+        String query = """
+        SELECT (
+            CASE 
+                WHEN DATEDIFF(t.date_created, t.transaction_date) > 7 THEN '7+ days'
+                ELSE CONCAT(DATEDIFF(t.date_created, t.transaction_date), ' days')
+            END
+        ) as days_backdated, COUNT(t.id) as shipments  FROM shipment s
+        INNER JOIN transaction t ON t.outgoing_shipment_id  = s.id
+        WHERE s.origin_id = :locationId 
+        AND t.date_created > :timeLimit
+        GROUP BY days_backdated
+        HAVING days_backdated > :daysOffset
+        """
+
+        return getBackdatedShipmentsChart(query, locationId, monthsLimit)
+    }
+
+    @Cacheable(value = "dashboardCache", key = { "getBackdatedInboundShipments-${locationId}${monthsLimit}" })
+    Map getBackdatedInboundShipmentsData(String locationId, Integer monthsLimit) {
+        String query = """
+        SELECT (
+            CASE 
+                WHEN DATEDIFF(t.date_created, t.transaction_date) > 7 THEN '7+ days'
+                ELSE CONCAT(DATEDIFF(t.date_created, t.transaction_date), ' days')
+            END
+        ) as days_backdated, COUNT(t.id) as shipments  FROM shipment s
+        INNER JOIN transaction t ON t.incoming_shipment_id  = s.id
+        WHERE s.destination_id = :locationId 
+        AND t.date_created > :timeLimit
+        GROUP BY days_backdated
+        HAVING days_backdated > :daysOffset
+        """
```

Comment:
the only difference between those two queries if origin vs destination. Can we somehow pass it as a variable here so that we can reuse the query in both methods?

Reply:
I am not sure if this is a good idea, I would rather say string concatenation in SQL is a thing that we should avoid.
As an example:
https://stackoverflow.com/questions/23179329/why-is-concatenating-sql-strings-a-bad-idea

If anyone else would like me to use the string concatenation here I can do it, but I am not a big fan, and that's the reason I didn't do it before.

Reply:
I know cases when concatenation is a wrong idea, and this is why they introduced an ability to use params via `:param` and pass it as the second argument of a query method, but in this case:
1. you wouldn't use any variable that the user provides and that might break the query (e.g. `drop database openboxes;` :laughing: ), but a variable that you would pass,
2. I was not really thinking about the concatenation, but rather passing it as param (if possible, probably not), or via `${variable}`.

I'm not sure if any of those options are possible, hence I just asked if "we can".